lib/elixir/lib/kernel/special_forms.ex

```diff
@@ -1408,6 +1408,32 @@ defmodule Kernel.SpecialForms do
   filters or inside the block, are not reflected outside of the
   comprehension.
 
+  Variable assignment based on `=` is treated as filters, and 
+  if the value of the expression is falsey (`nil` and `false`), it will be filtered out.
+  in case you don't want variable assignment to be treated as filters, 
+  you could move variable assignment into `do` block, or use a `<-` generator
+  and a single-element list literal on its right-hand side.
+      # A incorrect comprehension for getting language with grandparent.
+      # because `erlang` and `prolog` doesn't have grandparent, assigning `grandparent` is treated as false filter.
+      iex> languages = [elixir: :erlang, erlang: :prolog, prolog: nil]
+      iex> for {language, parent} <- languages, grandparent = languages[parent], do: {language, grandparent}
+      [elixir: :prolog]
+
+      # A correct comprehension for getting language with grandparent.
+      # Here, assignment is moved into `do` block, and not treated as filter.
+      iex> languages = [elixir: :erlang, erlang: :prolog, prolog: nil]
+      iex> for {language, parent} <- languages do
+      ...>   grandparent = languages[parent]
+      ...>   {language, grandparent}
+      ...> end
+      [elixir: :prolog, erlang: nil, prolog: nil]
+
+      # Alternative correct comprehension for getting language with grandparent.
+      # Here, assignment changed as generator (<-) with one element list, useful in case you have other filters which depend on `grandparent` data.
```

Comment:
```suggestion
      # Here, the assignment is converted to a generator (<-) with a single-element list, which is useful in case you have other filters which depend on the `grandparent` data.
```